Add multi-scale unwrapping

gmgunter opened the pull request:
This PR adds basic multi-scale unwrapping functionality.

### API updates

A new public function `multiscale_unwrap()` is added, which estimates unwrapped phase from an interferogram using a multi-resolution unwrapping strategy.

First, the interferogram is multilooked down to lower resolution and unwrapped to form a coarse estimate of the unwrapped phase. Then, the original interferogram is broken down into smaller tiles, which are each unwrapped independently. In order to mitigate phase discontinuities at tile boundaries, the phase values within each tile are then augmented by some number of phase cycles, using the coarse unwrapped estimate as a reference for the absolute phase. Specifically, the phase within each connected component is adjusted by a number of phase cycles in order to minimize its mean absolute difference with the low-resolution unwrapped phase estimate.

### Test cases

Two simple test cases have been added to exercise the new functionality.

The first test case simulates a noise-less interferogram with a roughly diagonal phase gradient such that, if the interferogram were unwrapped by tiles without any post-processing correction, there would be a significant phase offset between each tile. It then checks that the resulting unwrapped phase field differs from the simulated absolute phase only by a fixed integer number of phase cycles, with no phase discontinuities between tiles.

The figure below demonstrates the test data and compares the result of the multi-scale unwrapping approach with a naive tile-based unwrapping approach with no secondary unwrapping. The interferogram was partitioned into a 2x2 grid of tiles for unwrapping. In the simple tile-based unwrapping result, large phase discontinuities are visible along the tile boundaries.

![fig1](https://user-images.githubusercontent.com/12984092/184758098-eadfc270-9ee7-4b81-b513-7e6db44c6394.png)

The second test case is similar to the first, but increases the complexity by partitioning the interferogram into several isolated regions of high-coherence pixels, each of which span multiple tiles and are separated by noisy pixels. The test checks the unwrapped phase values within each high-coherence region independently. 

The figure below shows the test dataset unwrapped using SNAPHU. Note that the connected component labels show inconsistencies and mislabelings between tiles. This issue will be addressed in subsequent work.

![fig2](https://user-images.githubusercontent.com/12984092/184758725-393150f0-dc36-43da-ac6d-aa9a76bd0109.png)

### Known limitations

The following known issues will be addressed in subsequent PRs:

* Multi-scale unwrapping using PHASS is currently not working (ICU & SNAPHU are working and both are exercised by the test suite). The current behavior is that the code raises an exception saying that the unwrapped phase values are not congruent with the input wrapped phase.

   This seems to be due to a small number of pixels in the unwrapped phase that are filled with a large negative value (`-10000`) that is inconsistent with the surrounding phase values. The pixels are labeled as belonging to a connected component so they should be valid.

   I'll need to investigate this further.

* Currently, each tile is unwrapped in serial, one-by-one, but in the future we could unwrap them in parallel (perhaps using `dask`).

* As noted above, connected component labels from different tiles are inconsistent. I plan to address this in my next PR. For now, only the output unwrapped phase is reliable.

gmgunter commented:
PHASS support should be fixed now.

In one of the unit tests, there was an issue where I wasn't properly excluding all invalid pixels when checking the unwrapped phase values. There were just a few invalid phase values, which wasn't enough to cause the test assertion to fail in the case of ICU or SNAPHU. But PHASS seems to set invalid unwrapped phase values to `-10000`, causing the test to fail.

The test implementation should be fixed now, and the test suite has been updated to run with PHASS in addition to ICU & SNAPHU.

scottstanie commented on src/tophu/multiscale.py:
```diff
@@ -0,0 +1,465 @@
+        raise ValueError("downsample factor must be >= 1")
+    if any(map(lambda n: n < 1, ntiles)):
+        raise ValueError("number of tiles must be >= 1")
+
```
What do you think of the idea of having something like this at this stage:
```python
    if np.prod(ntiles) == 1:
        return unwrap(
            igram=igram,
            corrcoef=coherence,
            nlooks=nlooks,
        )
```

since right now, if you try and do `tiles = (1, 1)`, it'll just filter it and unwrap, then adjust components from filtered to unfiltered version.

(I actually can't get the tile=(1, 1) case to run, since it tells me the `bandwidth` of the filter is > 1 and invalid)

gmgunter replied:
Sounds good, I like the idea of special-casing the `ntiles=(1, 1)` case to avoid a bunch of unnecessary steps. I hadn't really considered that use case at all.

>(I actually can't get the tile=(1, 1) case to run, since it tells me the bandwidth of the filter is > 1 and invalid)

Good catch, seems like there's an issue when `downsample_factor` is either `(N, 1)` or `(1, N)`, unless you mess with the filter parameters. In this case, the multilooking & upsampling steps are no-ops along at least one axis, but the low-pass filtering is still performed, which raises an exception with the default filter parameters. Probably, we should skip low-pass filtering for each axis where `downsample_factor` was `1`.

I'll implement the necessary changes and add a few more simple test cases.

scottstanie replied:
Oh i just thought of something about this actually-

Maybe the special case should only be if both `ntiles=(1,1)` *and* `downsample_factor=(1,1)` (so neither tiling nor downsampling).

The original version I saw of using downsampled SNAPHU to unwrap higher-res images didn't do tiling, but did run snaphu on extra-multilooked (like in https://www.sciencedirect.com/science/article/pii/S0034425720304338).
That seems like a very useful case to have as an option here.

gmgunter replied:
Sounds good to me, I think that makes sense. Thanks for the paper :raised_hands:

scottstanie reviewed (changes requested):
this is awesome, and will probably become my go-to wrapper to run one of the unwrappers on numpy arrays in python.

gmgunter commented on src/tophu/multiscale.py:
```diff
@@ -0,0 +1,465 @@
+        The shape factor of the filter (the ratio of the width of the combined
+        pass-band and transition band to the pass-band width). Must be greater than or
+        equal to 1. A larger shape factor results in a more gradual filter roll-off.
+        (default: 1.5)
```
Oops, thanks for testing this. I'll try to add a similar case to the unit test suite.

It looks like this is a result of a bug on line 76 where we're calculating `bandwidth`, which causes the filter pass-band to be much wider than intended.

https://github.com/opera-adt/tophu/blob/cf1996a26559dec0372e7a8525c41e37a198b4cf/src/tophu/multiscale.py#L76

I think I just got the arithmetic wrong here.

Let $B_p$ be the pass-band width, $B_{p+t}$ be the combined pass-band + transition band width, and let $R$ be the ratio of the new sampling rate after downsampling relative to the original sampling rate.

The `shape_factor` parameter, $S$, is defined as

$$
S = \frac{B_{p+t}}{B_p}
$$

The `overhang` parameter, $O$, is defined as

$$
O = \frac{B_{p+t} - R}{B_{p+t} - B_p}
$$

So, if you solve for $B_p$ (the `bandwidth` variable), I think you should get

$$
B_p = \frac{R}{S + O - SO}
$$

which is different from what I have implemented here :facepalm:. As a result, the filter pass-band is much larger than it should be, which is causing the `bandpass_equiripple_filter()` function to complain.

---

Regarding this check that's causing the exception to be raised:

https://github.com/opera-adt/tophu/blob/cf1996a26559dec0372e7a8525c41e37a198b4cf/src/tophu/filter.py#L109-L110

This is basically checking that the low-pass filter has a valid transition band and stop band.

If the shape factor is exactly equal to `max_shape`, then the bandwidth of the filter stop-band will be zero. In order words, the transition band will extend all the way out to the Nyquist frequency. Similarly, if the shape factor is equal to `1`, there will be no transition band between the pass-band & the stop-band.

Technically, I suppose you could argue that these edge cases are valid filters, but they clearly seem like pathological cases that aren't likely to be what the user wanted. So I think the check still seems reasonable to me as implemented.

>I now see that the error message strictly less than, but the check is less than or equal

I think you might've misread. The check seems to be consistent with the error message to me. If `shape >= max_shape`, an exception is raised. The error message is saying `"shape factor must be [...] < {max_shape}"`.

---

Here's what the frequency response of the filter looks like for your test case with the current (buggy) implementation:

![fig1](https://user-images.githubusercontent.com/12984092/185294536-1036ff70-b1b1-483e-951b-5072d2bfb18e.png)

And here's what it looks like the frequency response looks like with the corrected `bandwidth` calculation:

![fig2](https://user-images.githubusercontent.com/12984092/185294540-2d90f3e1-8a89-4c1f-84ef-26a84f39e060.png)

Clearly, in the first figure the pass-band is way too wide and the transition band extends all the way out to the Nyquist frequency so there's no discernible stop-band.

In the second figure, the pass-band and stop-band cutoff frequencies (indicated by the dashed vertical lines) are at `0.2` and `0.3` respectively, which is consistent with `downsample_factor=2` and the default `overhang` value.

[Notebook](https://github.com/gmgunter/tophu-notebooks/blob/main/Equiripple-Low-Pass-Filter.ipynb)

gmgunter commented on src/tophu/multiscale.py:
```diff
@@ -0,0 +1,465 @@
+            # Compute the number of 2pi cycles that should be removed from the
+            # high-resolution phase in order to minimize the mean difference between the
+            # high-res and low-res phase.
+            avg_offset = np.mean(
```
Thanks for the feedback, @scottstanie!

I think this is a good idea too. But, if it's OK with you, I'll punt on this for now and implement it in a subsequent PR.

gmgunter commented:
@scottstanie thanks again for the review. I think I've addressed all of your comments in the latest commits, excepting [this suggestion](https://github.com/opera-adt/tophu/pull/3#discussion_r947256099) that I plan to implement in a future PR.

hfattahi commented on src/tophu/multiscale.py:
```diff
@@ -0,0 +1,482 @@
+    attenuation: float = 40.0,
+) -> NDArray:
+    r"""
+    Apply an anti-aliasing pre-filter, then multilook.
```
Ok after this comment I looked at the multilook function and I noticed that it actually does average. Basically from the higher resolution to the low resolution, it looks like we are doing two filters and one decimation (note that I consider multilook itself as a low-pass filter with average kernel plus decimation). 
I think it is a great functionality built in and extra smoothing usually helps with unwrapping. However, is it configurable if the user wants to apply only multi-looking and avoid the extra filtering?
That would give us the opportunity to understand the impact of the extra filter.

gmgunter replied:
Thanks for the review @hfattahi!

You're correct -- the "multilooking" step does perform spatial averaging. That step is functionally equivalent to applying a uniform weight filter and then decimating, so overall this `lowpass_filter_and_multilook()` function is effectively filtering twice before decimating.

My reasoning for filtering twice is that the two filters are intended to accomplish two slightly different things. The initial low-pass filter is there to reduce aliasing effects, since the input interferogram may contain high-frequency content that will otherwise cause artifacts when the interferogram is downsampled. The goal of this filter is to suppress the high-frequency content that would be undersampled after decimating, while preserving the low-frequency components. The second (spatial averaging) filter is intended to improve SNR by taking the mean of nearby pixels.

There is some overlap in the functionality of the two filters -- the low-pass filter does beat down the noise power, and the spatial averaging filter does suppress the higher frequencies a bit. But I don't think that spatial averaging alone would be sufficient to avoid aliasing effects that would distort the phase gradients in the downsampled interferogram.

The figures below help demonstrate this point. They show the frequency response of each filter for different downsampling factors. The dashed vertical line indicates the where the new Nyquist rate would be after downsampling.

![figure1](https://user-images.githubusercontent.com/12984092/189554967-28e693d7-42a1-41fb-b779-67881340bf81.png)
![figure2](https://user-images.githubusercontent.com/12984092/189554976-9e277d5f-7288-4844-93ce-9404e1a7b6ba.png)

[Notebook](https://github.com/gmgunter/tophu-notebooks/blob/main/LowPass-vs-Spatial-Average-Response.ipynb)

>is it configurable if the user wants to apply only multi-looking and avoid the extra filtering?

Right now, the low-pass filter is always applied whenever the interferogram is multilooked. I'll add an option to let the user disable it. Should be a pretty trivial change.

hfattahi replied:
Thank you @gmgunter for the nice plots and detailed explanation. It makes sense and nicely shows how the low-pass filter suppresses frequencies higher than nyquist.
It will be really great to make the filter configurable so that we could evaluate the impact on simulated and real data.

hfattahi commented on src/tophu/multiscale.py:
```diff
@@ -0,0 +1,482 @@
+    Parameters
+    ----------
+    igram_hires : numpy.ndarray
+        The full-resolution interferogram. A two-dimensional array.
```
Is this complex interferogram or only phase?
```suggestion
        The full-resolution wrapped interferogram. A two-dimensional array.
```

gmgunter replied:
It's the complex-valued interferogram, not the wrapped phase. But actually in this function I think we only need the phase of the interferogram, not the interferogram itself. So I guess it would make sense to change this interface to take the wrapped phase instead.

hfattahi reviewed (dismissed):
Thank you @gmgunter for the nice implementation of this part of the algorithm. It all looks great to me. I had few comments on the low-pass filter + multilook. Perhaps a clarification on why we need that extra spatial filter would be nice. After writing the comments, now I remember that you built in that low-pass filter to reduce high frequency artifacts caused by the sharp boundary of the uniform boxcar filter of multilooking window. Is that right?
